File: 백준/Gold/16987. 계란으로 계란치기/계란으로 계란치기.cc
#include <iostream>
using namespace std;

int N, ans;
int health[8], attack[8];

void backTracking(int picked) {
    if(picked == N) {
        int broken = 0;
        for(int i=0; i<N; i++)
            if(health[i] <= 0) broken++;
        if(broken > ans) ans = broken;
        return;
    }
    if(health[picked] <= 0) return backTracking(picked + 1);
    
    bool hasCrashed = false;
    for(int selected = 0; selected < N; selected++) {
        if(selected == picked) continue;
        if(health[selected] <= 0) continue;
        health[selected] -= attack[picked];
        health[picked] -= attack[selected];
        backTracking(picked + 1);
        health[selected] += attack[picked];
        health[picked] += attack[selected];
        if(not hasCrashed) hasCrashed = true;
    }
    if(not hasCrashed) backTracking(picked + 1);
}

int main() {
    ios::sync_with_stdio(false); cin.tie(NULL);
    cin >> N;
    for(int i=0; i<N; i++) 
        cin >> health[i] >> attack[i];
    backTracking(0);
    cout << ans;
}

File: 백준/Gold/16987. 계란으로 계란치기/계란으로 계란치기.py
import sys
readline = sys.stdin.readline

N = int(readline())
eggs = [list(map(int, readline().split())) for _ in range(N)]

def backTracking(picked):
  global ans
  if picked == N:
    cnt = 0
    for hp, _ in eggs:
      if hp <= 0:
        cnt += 1
    ans = max(ans, cnt)
    return
  
  if eggs[picked][0] <= 0:
    return backTracking(picked + 1)
  
  hasCrashed = False
  for selected in range(N):
    if selected == picked:
      continue
    if eggs[selected][0] <= 0:
      continue
    eggs[selected][0] -= eggs[picked][1]
    eggs[picked][0] -= eggs[selected][1]
    backTracking(picked + 1)
    eggs[selected][0] += eggs[picked][1]
    eggs[picked][0] += eggs[selected][1]
    if not hasCrashed:
      hasCrashed = True
  
  if not hasCrashed:
    backTracking(picked + 1)

ans = 0
backTracking(0)
print(ans)